recipes/qat_lora_finetune_distributed.py

```diff
@@ -0,0 +1,973 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import sys
+import time
+
+from functools import partial
+from typing import Any, Dict, List, Optional, Tuple, Union
+from warnings import warn
+
+import torch
+from omegaconf import DictConfig, ListConfig
+
+from torch import nn
+from torch.distributed import destroy_process_group, init_process_group
+
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader, DistributedSampler
+from torchtune import config, modules, training, utils
+from torchtune.config._utils import _get_component_from_path
+from torchtune.data import padded_collate_packed
+from torchtune.datasets import ConcatDataset
+from torchtune.modules.peft import (
+    DoRALinear,
+    get_adapter_params,
+    get_adapter_state_dict,
+    get_lora_module_names,
+    get_merged_lora_ckpt,
+    LoRALinear,
+    set_trainable_params,
+    validate_missing_and_unexpected_for_lora,
+)
+from torchtune.recipe_interfaces import FTRecipeInterface
+from torchtune.training import DummyProfiler, PROFILER_KEY
+from torchtune.training.quantization import (
+    _swap_lora_linear_with_qat,
+    _torchao_0_7_supported,
```

Comment:
In general we try not to have private APIs directly in our recipes like this. Personally I would just make `swap_lora_linear_with_qat` public and provide a bit more color in the docstring (though you can also add the caveat that it is subject to change). For `_torchao_0_7_supported` it probably doesn't merit a public API, but maybe can just repeat the try/except import at the beginning of the recipe. (Btw what I actually want us to do is provide extra metadata in the recipe registry around version requirements for each recipe, but unfortunately haven't gotten around to doing that yet.)